Split-attrs: Cube metadata refactortests

pp-mo opened the pull request:
A preliminary to addressing #4982 
I discovered that the CubeMetadata testing really needed a refactor, before I could attempt adding a 'global_attributes' field.
So this is that.

Using Pytest parametrization allows fewer tests and (a bit) less boilerplate, while also testing **_all_** the fields (which was not the case before).
Some of the behaviours this now highlights may be questionable, but I'm going to ruthlessly ignore that here AFAP (i.e. unless something breaks).

The attributes behaviour still uses extra tests, but I think that is quite reasonable.  
In future, I hope to further parametrize the attribute tests, to cover both `global_attributes` and ("local")`attributes`

pp-mo commented:
NOTE: rebased from #4981 branch onto "bare" existing feature-branch

Since the changes are entirely independent, and this may well go in first, now.

lbdreyer reviewed (changes requested):
The reduction in boilerplate is a big improvement and makes it much nicer to see what's going on with the tests!
I noticed there don't appear to be any tests for Test_difference.test_op_different_value but there are testing for that in Test__eq__ and Test_combine. Is there a reason for this?

pp-mo commented on lib/iris/tests/unit/common/metadata/test_CubeMetadata.py:
```diff
@@ -123,107 +138,114 @@ def test_call(self):
+            assert rmetadata.__eq__(lmetadata) == expect_success
+
+    def test_op_different__attribute_value(self, op_leniency):
+        # Check when one set of attributes has an extra entry.
```
:+1:

pp-mo commented on lib/iris/tests/unit/common/metadata/test_CubeMetadata.py:
```diff
@@ -472,178 +503,124 @@ def test_lenient(self):
-        rexpected = deepcopy(self.none)._asdict()
-        rexpected["cell_methods"] = lexpected["cell_methods"][::-1]
+    def test_op_different__none(self, fieldname, op_leniency):
+        if fieldname in ("attributes",):  # 'units'):
```
:+1:

pp-mo commented:
Thanks @lbdreyer I hope this addresses your points -- please re-review !
I have added a `Test_difference.test_op_different_value` as you suggest. Also rationalised the comment descriptions of similar testcases.

lbdreyer reviewed (approved):
LGTM!
